### PR TITLE
Exclude background task execution from root server span in ASGI middleware 

### LINKED ISSUE #1684: Exclude background tasks from request duration in ASGI middleware
**Is your feature request related to a problem?**
The duration of a request span in FastAPI includes the duration of evaluating background tasks. There is no span whose duration represents the time bewteen receiving a request and sending the response. An example:

```python
import asyncio
from typing import TypedDict

from fastapi import BackgroundTasks, Depends, FastAPI
from opentelemetry import trace
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter

app = FastAPI()
exporter = ConsoleSpanExporter()
provider = TracerProvider()
processor = BatchSpanProcessor(exporter)
provider.add_span_processor(processor)
trace.set_tracer_provider(provider)
FastAPIInstrumentor.instrument_app(app)


def get_context(
    background_tasks: BackgroundTasks,
):
    return Context(background_tasks=background_tasks)


async def some_background_task():
    await asyncio.sleep(10)


class Context(TypedDict):
    background_tasks: BackgroundTasks


@app.get("/")
async def main(context=Depends(get_context)):
    print(context.keys())
    context["background_tasks"].add_task(some_background_task)
    return None

```

**Describe the solution you'd like**
Either end the server span after we have sent the response (this is how dd-trace solve it in https://github.com/DataDog/dd-trace-py/pull/3799)

*OR* create a new span that stops after the response has been sent.

**Additional context**
This is important to calculation latency metrics using the FastAPI instrumentation

### PR DESCRIPTION
**Pull request by siminn-arnorgj**
# Description

Some ASGI web frameworks, like FastAPI, support scheduling of background tasks that execute after the server has sent a response to the client. Currently, the root server span doesn't end when the response is dispatched but rather when the server has completely processed the request. This can skew aggregate statistics, potentially making it appear as if the server has high latency when, in reality, it's only executing background tasks that don't impact the actual latency.

To address this, I've made changes to end the `server_span` upon sending the final response body. To ensure there are no duplicate attempts at ending the `server_span`, the span is initialized with `end_on_exit=False` and then ended manually if it's still recording.

**However**, one challenge with my solution pertains to error reporting. In the original implementation, exceptions in background tasks get reported on the top-level server span. With the changes in this PR, errors won't be reported unless a tracing context is active within the background task. Perhaps the FastAPI instrumentation should introduce a broader span for background task execution to ensure errors are reported within some span.

The solution is heavily inspired by the Datadog PR for this exact same problem https://github.com/DataDog/dd-trace-py/pull/3799

Before
image

After
image

Fixes #1684

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- [X] Ran in production for months (Python 3.11 FastAPI)
- [X] Manually tested with FastApi(0.103.1) in matrix [Normal Request, Request with background execution] x [No Exceptions, Exception before response, Exception in background task*]
- [X] Ran ASGI instrumentation tests
- [X] Ran FastAPI instrumentation tests
- [X] Added additional tests to the ASGI instrumentation

# Does This PR Require a Core Repo Change?

- [X] No.

# Checklist:

- [x] Followed the style guidelines of this project
- [X] Changelogs have been updated
- [X] Unit tests have been added
- [ ] ~~Documentation has been updated~~

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py:159]: Is this a common idiom for background work in ASGI? The user could get the behavior this PR is adding today by using `asyncio.create_task()`.

**Review by aabmass — COMMENTED** [review]
Thanks for the PR. Do you have any idea on the current state of OpenTelemetry [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.m) around this? I remember some effort to standardize when an HTTP span should be ended, but idk if it was for client or server.

File: instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py
Comment (aabmass) [reply instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py:159]: Thanks for the thorough answer. If this complies with semconv generally, it makes sense to me. The behavior may be slightly surprising to some users but I think it is semantically more meaningful.

Another option would be using span events to indicate when writing the body is complete without ending the span

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:698]: Small thing which a coworker pointed out; the request event could indicate that trailers are expected after the request body: https://asgi.readthedocs.io/en/latest/specs/www.html#response-start-send-event:~:text=If%20more_body%20is%20set%20to%20False%2C%20and%20the%20server%20is%20not%20expecting%20Response%20Trailers%20this%20will%20complete%20the%20response. In that case, I think you would want to wait for trailers to be sent before ending the span

File: instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py
Comment (aabmass) [reply instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py:159]: I think you've looked harder at the HTTP semconv than I have so LGTM

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py
Comment (siminn-arnorgj) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py]: In my latest commit(6a2ff51) I added support for HTTP trailers according to the ASGI spec. The drawback of supporting this is of course having to maintain state throughout the request. I currently maintain the state through a nonlocal variable. I'm open to suggestions on more elegant ways to track the state as I don't find the current method very elegant.

However I do think some discussion is warranted around HTTP trailers because AFAIK there is no publicly available ASGI server nor framework which conforms to the spec / supports HTTP trailers. So I'm not exactly sure the instrumentation should support something that nobody seems to use at the cost of increasing complexity.

Comment (aabmass) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py]: The closure solution looks pretty clean to me, I don't have any issue with it. I agree this may be unimportant in practice, but nice to have for completeness since it's only a few extra lines

**Review by aabmass — APPROVED** [review]
Thanks for the PR, LGTM and I think this is the right direction

**Comment by aabmass** [comment]
@open-telemetry/opentelemetry-python-contrib-maintainers can someone take a look and merge if it looks good?

**Comment by lzchen** [comment]
@siminn-arnorgj 

Would you mind rebasing with the latest main?

**Review by lzchen — APPROVED** [review]

**Pull request merged**
